src/btree/bt_compact.c

```diff
@@ -391,6 +391,22 @@ __wt_compact(WT_SESSION_IMPL *session)
             WT_WITH_PAGE_INDEX(session, ret = __compact_walk_internal(session, ref));
         }
 
+        if (ret == EBUSY) {
+            WT_STAT_CONN_INCR(session, session_table_compact_conflicting_checkpoint);
+
+            __wt_verbose_level(session, WT_VERB_COMPACT, WT_VERBOSE_DEBUG_1,
+              "The compaction of the data handle %s returned EBUSY due to an in-progress "
+              "conflicting checkpoint.%s",
+              session->dhandle->name,
+              session == S2C(session)->background_compact.session ?
+                "" :
+                " Compaction of this data handle will resume after checkpoint completes.");
+
+            __wt_sleep(1, 0);
+
+            ret = 0;
```

Comment:
This sets `ret` to 0 regardless of whether this is background or foreground compaction. Is this intentional?

Reply:
Yes, @etienneptl and I discussed this and we're concerned that for large tables background compact might never be able to correctly reclaim space if it continually gets interrupted by checkpoint.

Reply:
@pmacko86, I'm yet to refactor to address Keiths point but to clarify a couple of your points:
`compact_state` is set to`WT_COMPACT_SUCCESS` if compact finds any work to do, which could be finding pages to rewrite, dirty pages, or deleted pages. It's set according to the somewhat confusing skipp pointer that gets passed around bt_compact.c and block_compact.c. If we ever decide not to skip something, we set `WT_COMPACT_SUCCESS`. The "another pass" behaviour is largely unchanged from what it was previously. Setting `ret = 0` here means we will continue the walk in this loop and not return in the `WT_ERR` macro below.